### ISSUE TITLE
using not rotationally symmetrical psf from file

### ISSUE THREAD

**Issue by BertoltLang** [IvS-KULeuven/PlatoSim3#407]
Hello,
this is a question concerning the use of mapped PSFs from file.

At the moment PlatoSim selects a PSF only on the basis of the stars radial distance to the optical axis. The PSF is then rotated around the rotation angle, which is calculated from the stars position on the CCD and the angle attribute from the chosen PSF file.

For some calculations this approach has its limits, since some specific worst-case PSF maps we want to use are not rotationally symmetrical. That means a selection of a PSF on the basis of only its distance to the optical axis is not sufficient. 

Would it be possible to implement a solution where we could use a .hdf5 file with PSF data which are not specifically ordered or named but have their position on the CCD (in Cartesian or Polar coordinates) and the focal length used for their creation as attributes?

For clarification I could provide a file like this via email. 

Thanks in advance

Bert

**Comment by JorisDeRidder** [comment]
Could you elaborate on the application, i.e. on what would you exactly want to test/verify with rotationally asymmetric PSFs?  Would it not suffice to have different PSF map files to investigate the different azimuths separately?

**Comment by s-geiser** [comment]
To jump on the question - As far as I see the given hdf5 psf file already has all the psf's given for 4 different azimuth values (0/90/180/270), currently all being identical as PlatoSim uses azimuth 0 per default.

Was this intended for a possible rotational asymmetry as mentioned by BertoldLang?

If so one could potentially pass on x/y variables instead of the radial distance when calling psf -> select to enable choosing an appropriate psf based on the general position on the ccd?

**Comment by denisGrBach** [comment]
A quick comment from my side. The assumption that the PSF is symmetrical only holds for nominal TOU designs. The real TOU will suffer from manufacturing- and misalignment errors of the lenses. Leonardo did a Monte Carlo analysis to quantify this and delivered some worst case design. We already used these designs and showed that the impact on the FGS is non negligible (up to factor 2 worse). To do so, we used a guerilla implementation of Bertolt but it would be neat to have a general approach for the PlatoSim master. I think a general way could be to store as much PSF realizations as needed with its angular position (polar or Cartesian) as attributes and choose the nearest neighbor for the simulation. For compatibility the diagonal approach should be keept as well. I think it would be useful to keep the temperature group to allow for different types of stars.

**Comment by carstenpaproth** [comment]
Just for illustration how a PSF with a constant radial distance to the optical axis could look like for different azimuths.
![psfradial](https://user-images.githubusercontent.com/26323228/70702154-ed1a0400-1ccd-11ea-99a0-77603cacdd64.gif)

**Comment by rhuygen** [comment]
The original PSFs were calculated by Demetrio and provided to us by LESIA (Réza Samadi). I don't know why the format was chosen, probably because they were rotational symmetric. 

To support non-rotational symmetric PSFs, the selection algorithm will need some changes and the HDF5 file structure will need to be defined.

**Comment by SaraRegibo** [comment]
@denisGrBach We never asked for the file that @BertoltLang proposed to send us for clarification...  Do you have such a file?

**Comment by carstenpaproth** [comment]
Here you are, one of Bertolt's PSF files:

blueRealPSF.hdf5 (30 Tage gültig)<https://dlrmax.dlr.de/get/74d8f6a9-535c-510b-a92b-09185cf90256/>

Von: Sara Regibo [mailto:notifications@github.com]
Gesendet: Dienstag, 3. März 2020 09:44
An: IvS-KULeuven/PlatoSim3
Cc: Paproth, Carsten; Comment
Betreff: Re: [IvS-KULeuven/PlatoSim3] using not rotationally symmetrical psf from file (#407)


@denisGrBach<https://github.com/denisGrBach> We never asked for the file that @BertoltLang<https://github.com/BertoltLang> proposed to send us for clarification... Do you have such a file?

-

**Comment by SaraRegibo** [comment]
Thanks @carstenpaproth! I see the following attributes in each of the groups:
- (`starPointing1`, `starPointing2`)
- `strehlratio`
(`centerCoordinates1`, `centerCoordinates2`)
(`centroidCoordinates1`, `centroidCoordinates2`)
- (`centroidOffset1`, `centroidOffset2`)

Could you, please, comment on those (to rule out misinterpretations from our side): what they mean, the units,...?  Once we know for sure which data to use to select the correct PSF, we're good to go :)  Thanks in advance!

Also, I see the PSFs are 1024 x 1024 sub-pixels.   Can you confirm that this corresponds to 8x8 pixels?

**Comment by carstenpaproth** [comment]
@SaraRegibo 

These PSFs were created by Martin Pertenais with Zemax (Huygens PSF), for every guide star a dedicated one. Bertolt put them into an hdf5 file and copied some of the parameters from the Zemax PSF files into the hdf5 file. I found these Zemax PSF files on my computer, so I can comment on those parameters.

- starPointing is the X and Y direction in deg
- centerCoordinates is the X and Y position of the center point on the FPA in millimeters
- the center point is at subpixel 513, 513, I think it is counted from 1
- the PSFs were calculated for 8x8 pixels with 128x128 subpixels per pixel, so 1024x1024 subpixels total
- centroidCoordinates is the X and Y position of the centroid on the FPA in millimeters
- centroidOffset is the difference between centerCoordinates and centroidCoordinates
- the Strehl ratio is a measure of the optical quality, for PLATO it indicates that we are far from diffraction limited, and thus there are aberrations and the Huygens PSF is possibly not a great model for the PLATO PSF

**Comment by SaraRegibo** [comment]
Procedure to follow in case we are using this kind of pre-computed PSFs:
- determine the focal-plane coordinates of the centre of the sub-field [mm] -> `Detector::getFocalPlaneCoordinatesOfSubfieldCenter()`
- loop over all groups in the HDF5 file (keep on moving to the next one, until an exception is thrown)
     - extract the attributes denoting the position of the PSF on the focal plane -> (`centerCoordinates1`, `centerCoordinates2`)
     - calculate the distance from the centre of the sub-field (calculated before)
     - remember for which group this distance is smallest -> this is the PSF we need
- account for the rotation if the CCD w.r.t. focal plane

---

- [x] determine how we can "know" which kind of pre-computed PSFs we are using -> extra configuration parameter? based on the content of the HDF5 file (the groups that are present)?

**Comment by SaraRegibo** [comment]
@carstenpaproth Could you check whether the implementation in pull request #438 works as desired, please?  Just comment on this issue with the outcome.  If you are happy with the changes, I'll merge the code into the `develop` branch as soon as possible.

To use the rotationally asymmetrical PSF, use the following configuration:

```
sim["PSF/Model"] = "MappedFromFileAsymmetrical"
sim["PSF/MappedFromFileAsymmetrical/Filename"] = <path to blueRealPSF.hdf5>
```

The other parameters in the `PSF/MappedFromFileAsymmetrical` block (`NumberOfPixels`, `ChargeDiffusionStrength`, `IncludeChargeDiffusion`, and `IncludeJitterSmoothing`) are the same as in the `PSF/MappedFromFileSymmetrical` block.

**Comment by SaraRegibo** [comment]
To avoid further delay, I have merged the pull request...

**Comment by carstenpaproth** [comment]
@SaraRegibo 

Thanks for implementing this feature!

I made a quick test and it didn't work as expected.
I simulated an arbitrary star and the PSF in the output hdf5 file is just the first PSF from the provided input PSF file. I expected that the nearest PSF would be selected.

**Comment by SaraRegibo** [comment]
Alright, I'll have another look tomorrow!

**Comment by SaraRegibo** [comment]
Apparently the attributes `centerCoordinates1` and `centerCoordinates2` are read incorrectly from the PSF file (all zero, instead of the true value).  Re-opening #437.

**Comment by SaraRegibo** [comment]
The attributes `centerCoordinates1` and `centerCoordinates2` are now correctly read from the PSF file (see #437), so this issue should be solved, too.  @carstenpaproth, could you check, please?  Thanks!

**Comment by carstenpaproth** [comment]
@SaraRegibo 

Thanks, it works and the nearest PSF gets selected. But I noticed that the output PSF seems to be slightly rotated compared to the input PSF.
Do you rotate the PSF?
I think, it is not necessary to rotate the asymmetrical PSFs, because there is no rotational symmetry. On the other hand, a rotation introduces more variance to the PSFs of different stars, which could be desirable.
Maybe, an additional option in the yaml file could allow to rotate or not.
What do you think?

**Comment by SaraRegibo** [comment]
@carstenpaproth In the `setPsfForSubfield()` method in the `DetectorWithAsymmetricalMappedPSF` class, the following is done:
* based on the focal-plane coordinates of the centre of the sub-field, the PSF is selected (as described above);
* the PSF is rotated, so its **orientation matches that of the CCD, rather than that of the focal plane**;
* rebinning from sub-pixels to pixels;
* further preparations for the actual convolution.

---

Regarding the rotation (step 2):

A rotation over `-orientationAngle` is performed, where `orientationAngle` is read from:
* `CCD/Orientation`, in case of a `Custom` CCD position
* `CCDPositions/Orientation`, in case of CCD 1/2/3/4.

The conversion from degrees (as in the configuration file) to radians, combined with limitations in the `rotate()` methods, may cause the slight difference you noticed (e.g. a rotation over 90° will be actually end up being a rotation over just a bit more/less than 90°).

**Comment by SaraRegibo** [comment]
Hi @carstenpaproth, could you check whether the orientation of the PSFs is better in pull request #449?  You can do this as follows:

```
git checkout -b SaraRegibo-rotation develop
git pull https://github.com/SaraRegibo/PlatoSim3.git rotation
```
and then do your tests in the `rotation` branch.

**Comment by carstenpaproth** [comment]
@SaraRegibo 

Thanks, I think it is working now as expected.

**Comment by SaraRegibo** [comment]
Merged #449.
